src/Components/Components/src/RenderTree/StackObjectPool.cs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Components.RenderTree
+{
+    // This is a very simple object pool that requires Get and Return calls to be
+    // balanced as in a stack. It retains up to 'maxPreservedItems' instances in
+    // memory, then for any further requests it supplies untracked instances.
+
+    internal class StackObjectPool<T> where T : class, new()
+    {
+        private readonly int _maxPreservedItems;
+        private readonly T[] _contents;
+        private int _numSuppliedItems;
+        private int _numTrackedItems;
+
+        public StackObjectPool(int maxPreservedItems)
+        {
+            _maxPreservedItems = maxPreservedItems;
+            _contents = new T[_maxPreservedItems];
+        }
+
+        public T Get()
+        {
+            _numSuppliedItems++;
+
+            if (_numSuppliedItems <= _maxPreservedItems)
+            {
+                if (_numTrackedItems < _numSuppliedItems)
+                {
+                    // Need to allocate a new one
+                    var newItem = new T();
```

Comment:
Be aware that this won't be super fast. `new T()` is actually `Activator.CreateInstance()`. If this class doesn't need to be generic, then consider doing a coupled thing.

Reply:
Good tip - thanks. I had previously assumed otherwise, but having read up on it, I see why now.

I've changed `StackObjectPool` to require you to pass an instance factory (i.e., `Func<T>`) to its constructor.

Reply:
TBH it wouldn't have mattered in our use cases. The whole point of these instances being pooled is that the cost of creating/collecting them amortises to zero if your app keeps running.